dlmalloc/src/malloc.c

```diff
@@ -4105,7 +4105,15 @@ static void* sys_alloc(mstate m, size_t nb) {
     msegmentptr ss = (m->top == 0)? 0 : segment_holding(m, (char*)m->top);
     ACQUIRE_MALLOC_GLOBAL_LOCK();
 
+#ifdef __wasilibc_unmodified_upstream
     if (ss == 0) {  /* First time through or recovery */
+#else
+    // In wasi-libc, this branch should never be taken as CALL_MORECORE(0) may
+    // lead to incorrect allocations. See try_init_allocator() for more
+    // information.
```

Comment:
this is inside `if (MORECORE_CONTIGUOUS && ...` block.
if we are using MORECORE_CONTIGUOUS=1, it's the bug to fix i guess.

Reply:
In general I would say setting `MORECORE_CONTIGUOUS` to 1 is correct. The comments in `malloc.c` say this:
```
MORECORE_CONTIGUOUS       default: 1 (true) if HAVE_MORECORE
  If true, take advantage of fact that consecutive calls to MORECORE
  with positive arguments always return contiguous increasing
  addresses.  This is true of unix sbrk. It does not hurt too much to
  set it true anyway, since malloc copes with non-contiguities.
  Setting it false when definitely non-contiguous saves time
  and possibly wasted space it would take to discover this though.
```
So, in general I think we definitely want to benefit from the fact that the WebAssembly memory is contiguous and only fall back to the non-contiguous malloc if some other allocator uses memory.grow in the same program, as well.

Reply:
ok. it makes sense.
i vaguely concern that discontiguous fallback logic will never be tested well though.